internal/pkg/manifest/pipeline.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"text/template"
+
+	"github.com/fatih/structs"
+	"gopkg.in/yaml.v3"
+
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/archer"
+	"github.com/aws/amazon-ecs-cli-v2/templates"
+)
+
+// Provider defines a source of the artifacts
+// that will be built and deployed via a pipeline
+type Provider interface {
+	fmt.Stringer
+	Name() string
+	Properties() map[string]interface{}
+}
+
+type githubProvider struct {
+	properties *GithubProperties
+}
+
+func (p *githubProvider) Name() string {
+	return "github"
+}
+
+func (p *githubProvider) String() string {
+	return "github"
+}
+
+func (p *githubProvider) Properties() map[string]interface{} {
+	return structs.Map(p.properties)
+}
+
+// GithubProperties contain information for configuring a Github
+// source provider.
+type GithubProperties struct {
+	// use tag from https://godoc.org/github.com/fatih/structs#example-Map--Tags
+	// to specify the name of the field in the output properties
+	Repository string `structs:"repository" yaml:"repository"`
+	Branch     string `structs:"branch" yaml:"branch"`
+}
+
+// NewProvider creates a source provider based on the type of
+// the provided provider-specific configurations
+func NewProvider(configs interface{}) (Provider, error) {
+	switch props := configs.(type) {
+	case *GithubProperties:
+		return &githubProvider{
+			properties: props,
+		}, nil
+	default:
+		return nil, &ErrUnknownProvider{unknownProviderProperties: props}
+	}
+}
+
+// PipelineSchemaMajorVersion is the major version number
+// of the pipeline manifest schema
+type PipelineSchemaMajorVersion int
+
+const (
+	// Ver1 is the current schema major version of the pipeline.yml file.
+	Ver1 PipelineSchemaMajorVersion = iota + 1
+)
+
+// PipelineManifest contains information that defines the relationship
+// and deployment ordering of your environments.
+type PipelineManifest struct {
+	Version      PipelineSchemaMajorVersion `yaml:"version"`
+	Source       *Source                    `yaml:"source"`
+	Environments []PipelineStage            `yaml:"stages"`
+}
+
+// Source defines the source of the artifacts to be built and deployed.
+type Source struct {
+	ProviderName string                 `yaml:"provider"`
+	Properties   map[string]interface{} `yaml:"properties"`
+}
+
+// PipelineStage represents configuration for each deployment stage
+// of an application.
+type PipelineStage struct {
+	Name string `yaml:"env"`
+}
+
+// CreatePipeline returns a pipeline manifest object.
+func CreatePipeline(provider Provider, stages ...PipelineStage) (archer.Manifest, error) {
+	// TODO: #221 Do more validations
+	var defaultStages []PipelineStage
+	if len(stages) == 0 {
+		defaultStages = []PipelineStage{
```

Comment:
It's cool to keep this now, but I think in the future, we'll just bail if they don't have any environments.